### DISCUSSION TITLE
MD240128A6W-FPTLW SPI 4W Support

### DISCUSSION THREAD

**Discussion by MarekOravec** [olikraus/u8g2#1482] [Q&A]
Hello guys, I have MD240128A6W-FPTLW 240x128 with ST7586S.

Is this display supported ? I tried every ST7586S library in u8g2 and not working...

My connection (I hope its good | for 3W/4W):
1 - Vdd 
3 - Vss
4 - Display CS
5 - Display Reset pin
6 - Vdd
13 - A0 signal
14 - SPI MOSI
15 - Vdd
16 - SPI CLK
17 - LED+
18 - LED-

Thanks for advice :)

**Comment by olikraus** [comment]
Did you test one the following constructors:
U8G2_ST7586S_ERC240160_1_4W_HW_SPI
U8G2_ST7586S_YMC240160_1_4W_HW_SPI

Especially test this with the ContrastTest (page buffer) example, because you may need to adjust contrast (setContrast()) for your display.

Of course wiring might be also an issue: Which Board? How are wires connected? What are your arguments to the constructor?

**Reply by MarekOravec** [reply]
Hi, i have ESP32dev board with SW SPI. On the back of the pcb display I set the mod (with resistors) to 3W SPI. I dont have U8G2_ST7586S_YMC240160_1_4W_HW_SPI (Identifier is undefined) maybe bad library ? Im using VSC with PlatformIO. I also looked at the connections and they should all be fine (tested multiple times).. Now i dont have A0 signal connected (pin 13), i see some pixels rendered after init but nothing else. I'll try something and write to you how it turned out... Maybe i need reconfigure display info in u8x8_d_st7586s_erc240160.c for my display

**Comment by olikraus** [comment]
I think both driver should at least display something. 4W SPI should be more reliable. For testing indeed I suggest 4W Software emulated SPI.

**Reply by MarekOravec** [reply]
Hi @olikraus , little update... I stay at 3W SPI and erc240160 3W SW SPI. I check datasheet and your code, column and row should be good (0x7F and 0x9F for row, for full scale). I edited pixel_width = 240, pixel_height = 128 cos i have 240x128 display. 

But always have this issue:
https://drive.google.com/file/d/13mOWXWpFBeHGp8q6X2Tmh8oKKnWNPdrf/view?usp=sharing

`  u8g2.setFont(u8g2_font_unifont_t_symbols);
  u8g2.setCursor(0,100);
  u8g2.print("hello world display init");
  u8g2.sendBuffer();`
  
  The font is stretched. The X and Y axes do not match the setting. It may just be a simple mistake, I did not study your code in detail. Can you help me with this?

**Reply by olikraus** [reply]
> edited pixel_width = 240, pixel_height = 128 

This will not fully work, because there is one more place to edit.

> The font is stretched. The X and Y axes do not match the setting. 

ST7586 displays are very complicated. I think in your case some reverse engineering is required to understand the layout / architecture of the display. Once this is known, we need to modify the data upload procedures:

https://github.com/olikraus/u8g2/blob/5efba243cbb317ed232e0e0017550faf4de78df9/csrc/u8x8_d_st7586s_erc240160.c#L186-L210 

Unfortunately, I am also not the author of this code, so the effort would be very hard for me to modify this.

Is there any detailed documentation/datasheet of your display?

**Reply by MarekOravec** [reply]
Hello, thanks for answer. 
I have only these two datasheets: http://www.farnell.com/datasheets/3164273.pdf , https://www.buydisplay.com/download/ic/ST7586.pdf

**Reply by olikraus** [reply]
Hmm this doesn't really help :(

**Reply by MarekOravec** [reply]
Hello @olikraus, It works well 4W SPI with U8G2_ST7586S_S028HN118A_F_4W_SW_SPI. So problem solved :)

**Reply by olikraus** [reply]
:)
